<a href="https://colab.research.google.com/github/sskamalapurkar/Unsupervised_ML_Clustering/blob/main/Online_Retail_Customer_Segmentation_Sayali_Kamalapurkar.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Problem Statement**

---



In this project, your task is to identify major customer segments on a transnational data set which contains all the transactions occurring between 01/12/2010 and 09/12/2011 for a UK-based and registered non-store online retail.The company mainly sells unique all-occasion gifts. Many customers of the company are wholesalers.

**In this project, we are required to :**


*   Understand the Dataset & cleanup (if required)

*   Build a clustering model to segment the customer-based similarity

*   Also fine-tune the hyperparameters & compare the evaluation metrics of various classification algorithms

# **Introduction**

---



In [22]:
#importing the libraries
import pandas as pd
import numpy as np
from numpy import math
import datetime as dt

#Visualization libraries
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline


#ignore warnings
import warnings
warnings.filterwarnings('ignore')

In [2]:
# importing dataset
file_path = '/content/drive/MyDrive/Unsupervised ML - Clustering/Online Retail.xlsx'
DataFrame = pd.read_excel(file_path)
df = DataFrame.copy()

In [3]:
# let us have a glance at the dataset
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

# **Data Description**

---

**Attribute Information**

|Field Name|Description|
|---|---|
|**InvoiceNo:**|Invoice number. A 6-digit integer number uniquely assigned to each transaction.
||If this code starts with letter 'c', it indicates a cancellation.|
|**StockCode:**|Product (item) code. A 5-digit integer number uniquely assigned to each distinct product.|
|**Description:**|Product (item) name|
|**Quantity:**|The quantities of each product (item) per transaction.|
|**InvoiceDate:**|Invice Date and time. The day and time when each transaction was generated.|
|**UnitPrice:**|Unit price. Product price per unit in sterling.|
|**CustomerID:**|Customer number. A 5-digit integral number uniquely assigned to each customer.|
|**Country:**|Country name. The name of the country where each customer resides.|

In [4]:
#Checking number of rows and columns
df.shape

(541909, 8)

In [5]:
print("Total retail transactions including each product in every transactions (Rows): ", df.shape[0])
print("NUmber of features (Columns): ", df.shape[1])

Total retail transactions including each product in every transactions (Rows):  541909
NUmber of features (Columns):  8


In [6]:
# printing names of all the columns / features
df.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')

In [7]:
# getting info about the columns
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


**There are 4 features whose datatype is object, 2 features whose datatype is float64, 1 feature whose datatype is datetime64 and 1 feature whose datatype is int64.**

**Memory usage by the dataset is 33.1 MB.**

In [8]:
#checking for unique values
df.nunique()

InvoiceNo      25900
StockCode       4070
Description     4223
Quantity         722
InvoiceDate    23260
UnitPrice       1630
CustomerID      4372
Country           38
dtype: int64



*   **For 4070 products, there are 25900 transactions in the data. This means that each product is likely to have multiple transactions in the data. There are almost as many products as customers in the data as well.**

*   **In the dataset, we have retail transactions from almost 38 different countries of the world.**

*   **Almost 25900 overall transactions, this includes cancelled transactions too.**

*   **Also, 4223 different items are listed in the dataset.**

*   **Overall customer count in the available dataset is 4372.**

In [9]:
#let us check for null values
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

**Only 2 columns have null values in it.**

# **Duplicate Values**

---

Why is it important to remove duplicate records from my data?

> "Duplication" means that we have repeated data in our dataset. This could be due to things like data entry errors or data collection methods. By removing duplicate values in our data set, time and money are saved by not sending identical communications multiple times to the same person.

In [10]:
# checking for duplicate Values in the dataset
duplicates=len(df[df.duplicated()])
print("The number of duplicate values in the dataset is - ",duplicates)

The number of duplicate values in the dataset is -  5268


In [11]:
df[df.duplicated()].head()

InvoiceNo StockCode                        Description  Quantity  \
517    536409     21866        UNION JACK FLAG LUGGAGE TAG         1   
527    536409     22866      HAND WARMER SCOTTY DOG DESIGN         1   
537    536409     22900    SET 2 TEA TOWELS I LOVE LONDON          1   
539    536409     22111       SCOTTIE DOG HOT WATER BOTTLE         1   
555    536412     22327  ROUND SNACK BOXES SET OF 4 SKULLS         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
517 2010-12-01 11:45:00       1.25     17908.0  United Kingdom  
527 2010-12-01 11:45:00       2.10     17908.0  United Kingdom  
537 2010-12-01 11:45:00       2.95     17908.0  United Kingdom  
539 2010-12-01 11:45:00       4.95     17908.0  United Kingdom  
555 2010-12-01 11:49:00       2.95     17920.0  United Kingdom

In [12]:
# Select transactions with InvoiceNo = '536409' & StockCode = '21866'
df.loc[(df['InvoiceNo'].astype(str) == '536409') & (df['StockCode'].astype(str) == '21866')]

InvoiceNo StockCode                  Description  Quantity  \
494    536409     21866  UNION JACK FLAG LUGGAGE TAG         1   
517    536409     21866  UNION JACK FLAG LUGGAGE TAG         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
494 2010-12-01 11:45:00       1.25     17908.0  United Kingdom  
517 2010-12-01 11:45:00       1.25     17908.0  United Kingdom

As we can see, the duplicates are nothing but repeat entry of the particular transactions. These entries are exactly same, means there is no significance of such entries while analyzing the data. Hence we will drop all those duplicate rows from the data.

In [13]:
# Remove duplicates
df = df.drop(index=df[df.duplicated()].index)
print("Number of rows after removal of duplicates : {}".format(df.shape[0]))

Number of rows after removal of duplicates : 536641


# **Data Preprocessing**

---



In [14]:
df.describe(include = 'all')

InvoiceNo StockCode                         Description  \
count    536641.0    536641                              535187   
unique    25900.0      4070                                4223   
top      573585.0    85123A  WHITE HANGING HEART T-LIGHT HOLDER   
freq       1114.0      2301                                2357   
first         NaN       NaN                                 NaN   
last          NaN       NaN                                 NaN   
mean          NaN       NaN                                 NaN   
std           NaN       NaN                                 NaN   
min           NaN       NaN                                 NaN   
25%           NaN       NaN                                 NaN   
50%           NaN       NaN                                 NaN   
75%           NaN       NaN                                 NaN   
max           NaN       NaN                                 NaN   

             Quantity          InvoiceDate      UnitPrice     CustomerID  \
count   536641.000000               536641  536641.000000  401604.000000   
unique            NaN                23260            NaN            NaN   
top               NaN  2011-10-31 14:41:00            NaN            NaN   
freq              NaN                 1114            NaN            NaN   
first             NaN  2010-12-01 08:26:00            NaN            NaN   
last              NaN  2011-12-09 12:50:00            NaN            NaN   
mean         9.620029                  NaN       4.632656   15281.160818   
std        219.130156                  NaN      97.233118    1714.006089   
min     -80995.000000                  NaN  -11062.060000   12346.000000   
25%          1.000000                  NaN       1.250000   13939.000000   
50%          3.000000                  NaN       2.080000   15145.000000   
75%         10.000000                  NaN       4.130000   16784.000000   
max      80995.000000                  NaN   38970.000000   18287.000000   

               Country  
count           536641  
unique              38  
top     United Kingdom  
freq            490300  
first              NaN  
last               NaN  
mean               NaN  
std                NaN  
min                NaN  
25%                NaN  
50%                NaN  
75%                NaN  
max                NaN

**Observations:**



*   There are 4070 unique product codes. **'85123A'** is the most common product code appeared 2301 times.

*   Majority of customers resides in United Kingdom.

In [15]:
# Inspect min and max Quantity transactions
df.loc[(df['Quantity']==-80995) | (df['Quantity']==80995)]

InvoiceNo StockCode                  Description  Quantity  \
540421    581483     23843  PAPER CRAFT , LITTLE BIRDIE     80995   
540422   C581484     23843  PAPER CRAFT , LITTLE BIRDIE    -80995   

               InvoiceDate  UnitPrice  CustomerID         Country  
540421 2011-12-09 09:15:00       2.08     16446.0  United Kingdom  
540422 2011-12-09 09:27:00       2.08     16446.0  United Kingdom

**Invoice No. 581483 of 80995 quantity sell of stock code 23843 is reversed by adding a contra entry. Invoice No. of such reversal/cancelled entries starts with letter 'C'**

In [16]:
# Cancelled/Reversed invoices
print("Cancelled invoices/transactions : {}".format(df[df['InvoiceNo'].astype(str).str[0] == 'C'].shape[0]))

Cancelled invoices/transactions : 9251


In [17]:
# Display transactions with negative unit price.
df[df['UnitPrice'] < 0]

InvoiceNo StockCode      Description  Quantity         InvoiceDate  \
299983   A563186         B  Adjust bad debt         1 2011-08-12 14:51:00   
299984   A563187         B  Adjust bad debt         1 2011-08-12 14:52:00   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

**Above transactions are looking to be a manual ledger adjustments. Also they don't have customer id.**

In [18]:
# Dsiplay transaction with unit price > 10000
df[df['UnitPrice'] > 10000]

InvoiceNo  StockCode      Description  Quantity         InvoiceDate  \
15016    C537630  AMAZONFEE       AMAZON FEE        -1 2010-12-07 15:04:00   
15017     537632  AMAZONFEE       AMAZON FEE         1 2010-12-07 15:08:00   
16232    C537644  AMAZONFEE       AMAZON FEE        -1 2010-12-07 15:34:00   
16356    C537651  AMAZONFEE       AMAZON FEE        -1 2010-12-07 15:49:00   
43702    C540117  AMAZONFEE       AMAZON FEE        -1 2011-01-05 09:55:00   
43703    C540118  AMAZONFEE       AMAZON FEE        -1 2011-01-05 09:57:00   
222681   C556445          M           Manual        -1 2011-06-10 15:31:00   
299982   A563185          B  Adjust bad debt         1 2011-08-12 14:50:00   
524601   C580604  AMAZONFEE       AMAZON FEE        -1 2011-12-05 11:35:00   
524602   C580605  AMAZONFEE       AMAZON FEE        -1 2011-12-05 11:36:00   

        UnitPrice  CustomerID         Country  
15016    13541.33         NaN  United Kingdom  
15017    13541.33         NaN  United Kingdom  
16232    13474.79         NaN  United Kingdom  
16356    13541.33         NaN  United Kingdom  
43702    16888.02         NaN  United Kingdom  
43703    16453.71         NaN  United Kingdom  
222681   38970.00     15098.0  United Kingdom  
299982   11062.06         NaN  United Kingdom  
524601   11586.50         NaN  United Kingdom  
524602   17836.46         NaN  United Kingdom

**Most of the higher unit price transactions are related to Amazon fee for selling the product through their platform. Such transactions don't have customer id.**

In [19]:
print("Transactions with zero unit price : {}".format(df[df['UnitPrice'] == 0].shape[0]))

Transactions with zero unit price : 2510


In [20]:
#First five transactions with zero unit price
df[df['UnitPrice'] == 0].head()

InvoiceNo StockCode Description  Quantity         InvoiceDate  UnitPrice  \
622     536414     22139         NaN        56 2010-12-01 11:52:00        0.0   
1970    536545     21134         NaN         1 2010-12-01 14:32:00        0.0   
1971    536546     22145         NaN         1 2010-12-01 14:33:00        0.0   
1972    536547     37509         NaN         1 2010-12-01 14:33:00        0.0   
1987    536549    85226A         NaN         1 2010-12-01 14:34:00        0.0   

      CustomerID         Country  
622          NaN  United Kingdom  
1970         NaN  United Kingdom  
1971         NaN  United Kingdom  
1972         NaN  United Kingdom  
1987         NaN  United Kingdom

We can see that such odd transactions do not have any customer ID. So it won't be useful to keep them in the data. It is advisable to drop all such columns which do not have customer ID.

In [21]:
#Remove transactions with missing customer ids
df = df.drop(index=df[df['CustomerID'].isnull()].index)
print("Retail transactions after removing missing customer ids  : {}".format(df.shape[0]))

Retail transactions after removing missing customer ids  : 401604


In [23]:
#InvoiceDate column has date and time, which we will separate as month, day and hour respectively
df['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'])

df['Month'] = df['InvoiceDate'].dt.month
df['day_name'] = df['InvoiceDate'].dt.day_name
df['Day'] = df['InvoiceDate'].dt.day
df['Hour'] = df['InvoiceDate'].dt.hour

In [24]:
# From quantity and unit price, we can calculate total sales amount for that perticular transaction
#df['Sales_Amount'] = df['Quantity'] * df['UnitPrice']